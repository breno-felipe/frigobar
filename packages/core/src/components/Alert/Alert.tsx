import React, { useEffect } from 'react';
import styled from 'styled-components';
import { useFade } from '@frigobar/animation';

const CloseIcon = ({ ...props }) => (
  <svg
    version="1.1"
    xmlns="http://www.w3.org/2000/svg"
    x="0px"
    y="0px"
    viewBox="0 0 47.971 47.971"
    {...props}
  >
    <g>
      <path
        d="M28.228,23.986L47.092,5.122c1.172-1.171,1.172-3.071,0-4.242c-1.172-1.172-3.07-1.172-4.242,0L23.986,19.744L5.121,0.88
		c-1.172-1.172-3.07-1.172-4.242,0c-1.172,1.171-1.172,3.071,0,4.242l18.865,18.864L0.879,42.85c-1.172,1.171-1.172,3.071,0,4.242
		C1.465,47.677,2.233,47.97,3,47.97s1.535-0.293,2.121-0.879l18.865-18.864L42.85,47.091c0.586,0.586,1.354,0.879,2.121,0.879
		s1.535-0.293,2.121-0.879c1.172-1.171,1.172-3.071,0-4.242L28.228,23.986z"
      />
    </g>
  </svg>
);

const CloseButton = styled.button<{ withText: boolean }>`
  align-items: center;
  background-color: transparent;
  border: 0;
  border-radius: 50%;
  cursor: pointer;
  display: inline-flex;
  justify-content: center;
  padding: 5px;
  position: absolute;
  right: 10px;
  transition: background-color 0.2s ease-in-out;
  top: 10px;

  &:hover,
  &:focus {
    background-color: rgba(0, 0, 0, 0.1);
  }

  ${({
    withText,
    theme: {
      colors: { neutral },
    },
  }): string | false =>
    withText &&
    `
      border-radius: none;
      color: ${neutral[700]};
      padding: 0;
      top: 12px;

      &:hover,
      &:focus {
        background-color: transparent;
        color: ${neutral[900]};
      }
    `}
`;

const Wrapper = styled.div<{ type: AlertProps['type'] }>`
  border-radius: 4px;
  font-size: 0.875rem;
  position: relative;
  ${({ theme: { colors, borders, spacings }, type = 'neutral' }): string => `
    background-color: ${colors[type][50]};
    border: ${borders.tiny}px solid ${colors[type][200]};
    color: ${colors.black};
    padding: ${spacings.small}px;
  `}
`;

interface CloseProps {
  onClick: (event: React.MouseEvent<HTMLElement>) => void;
  text?: string;
  ariaLabel?: string;
}

const Close = ({
  onClick,
  text,
  ariaLabel,
  ...props
}: CloseProps): JSX.Element => (
  <CloseButton
    {...props}
    onClick={onClick}
    withText={Boolean(text)}
    aria-label={text ? undefined : ariaLabel}
  >
    {!text ? <CloseIcon width="10" height="10" /> : <span>{text}</span>}
  </CloseButton>
);

Close.defaultProps = {
  text: undefined,
  ariaLabel: 'close',
};

interface AlertProps {
  /** content to be displayed inside alert */
  children: React.ReactNode;
  /** change the colors of the wrapper */
  type?: 'success' | 'info' | 'danger' | 'warning' | 'neutral';
  /** display close button */
  closable?: boolean;
  /** function when close button is clicked */
  onClose?: (event: React.MouseEvent<HTMLElement>) => void;
  /** if you want to display a text instead of the "X" button */
  closeText?: string;
  /** aria-label to be in close button */
  closeIconAriaLabel?: string;
  /** shows alert if true */
  show: boolean;
}

const Alert = ({
  children,
  type = 'neutral',
  closable,
  onClose,
  closeText,
  closeIconAriaLabel,
  show = false,
  ...props
}: AlertProps): JSX.Element | null => {
  const [{ animation, state }, toggle] = useFade({ startOnRender: show });

  useEffect(() => {
    toggle(show);
  }, [show, toggle]);

  return state ? (
    <Wrapper
      {...props}
      type={type}
      role="alert"
      css={`
        animation: ${animation};
      `}
    >
      {closable && (
        <Close
          text={closeText}
          ariaLabel={closeIconAriaLabel}
          onClick={(e): void => {
            onClose && onClose(e);
          }}
        />
      )}
      {children}
    </Wrapper>
  ) : null;
};

export default Alert;
