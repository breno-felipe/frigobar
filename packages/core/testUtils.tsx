import React from 'react';
import PropTypes from 'prop-types';
import {
  render,
  fireEvent,
  waitFor,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import theme from './src/theme';

const Providers = ({ children }: { children?: React.ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

Providers.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, fireEvent, waitFor };
