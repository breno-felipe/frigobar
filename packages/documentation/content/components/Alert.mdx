---
title: Alert
menu: components
---

Alert component

```jsx react-live=true
<Alert show type="danger">
  Something went wrong!
</Alert>
```
