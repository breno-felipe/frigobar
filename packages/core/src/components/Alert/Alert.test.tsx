import React from 'react';
import { render, fireEvent, waitFor } from '../../../testUtils';

import Alert from './Alert';

describe('<Alert />', () => {
  describe('Snapshots', () => {
    it('should match snapshot with all types', () => {
      const alerts = [
        <Alert show type="neutral" key="neutral">
          Alert
        </Alert>,
        <Alert show type="info" key="info">
          Alert
        </Alert>,
        <Alert show type="warning" key="warning">
          Alert
        </Alert>,
        <Alert show type="danger" key="danger">
          Alert
        </Alert>,
        <Alert show type="success" key="success">
          Alert
        </Alert>,
      ];

      alerts.forEach(Component => {
        const { container } = render(Component);
        expect(container).toMatchSnapshot();
      });
    });
    it('should match snapshot with closable', () => {
      const { container } = render(
        <Alert show closable>
          Alert
        </Alert>,
      );
      expect(container).toMatchSnapshot();
    });
    it('should match snapshot with closable and closeText', () => {
      const { container } = render(
        <Alert show closable closeText="Close Now">
          Alert
        </Alert>,
      );
      expect(container).toMatchSnapshot();
    });
  });
  describe('Closable feature', () => {
    it('should close alert when click on close button', async () => {
      const { getByText, queryByText } = render(
        <Alert show closable closeText="Close Now">
          Alert
        </Alert>,
      );

      expect(queryByText('Alert')).toBeTruthy();

      fireEvent.click(getByText('Close Now'));

      waitFor(() => {
        expect(queryByText('Alert')).toBeFalsy();
      });
    });
  });
  describe('onClose function', () => {
    it('should call onClose function when alert is closed', () => {
      const onCloseMock = jest.fn();
      const { getByText } = render(
        <Alert show closable closeText="Close Now" onClose={onCloseMock}>
          Alert
        </Alert>,
      );

      fireEvent.click(getByText('Close Now'));

      expect(onCloseMock).toHaveBeenCalled();
    });
  });
});
