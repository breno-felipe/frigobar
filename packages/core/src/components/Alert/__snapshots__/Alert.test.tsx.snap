// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> Snapshots should match snapshot with all types 1`] = `
.c0 {
  border-radius: 4px;
  font-size: 0.875rem;
  position: relative;
  background-color: #fdfdfd;
  border: 1px solid #ccc;
  color: #000;
  padding: 12px;
}

.c1 {
  -webkit-animation: lbWRkT 300ms forwards;
  animation: lbWRkT 300ms forwards;
}

<div>
  <div
    class="c0 c1"
    role="alert"
    type="neutral"
  >
    Alert
  </div>
</div>
`;

exports[`<Alert /> Snapshots should match snapshot with all types 2`] = `
.c0 {
  border-radius: 4px;
  font-size: 0.875rem;
  position: relative;
  background-color: #e5e7f0;
  border: 1px solid #929cbf;
  color: #000;
  padding: 12px;
}

.c1 {
  -webkit-animation: lbWRkT 300ms forwards;
  animation: lbWRkT 300ms forwards;
}

<div>
  <div
    class="c0 c1"
    role="alert"
    type="info"
  >
    Alert
  </div>
</div>
`;

exports[`<Alert /> Snapshots should match snapshot with all types 3`] = `
.c0 {
  border-radius: 4px;
  font-size: 0.875rem;
  position: relative;
  background-color: #f9f2e9;
  border: 1px solid #e6caa3;
  color: #000;
  padding: 12px;
}

.c1 {
  -webkit-animation: lbWRkT 300ms forwards;
  animation: lbWRkT 300ms forwards;
}

<div>
  <div
    class="c0 c1"
    role="alert"
    type="warning"
  >
    Alert
  </div>
</div>
`;

exports[`<Alert /> Snapshots should match snapshot with all types 4`] = `
.c0 {
  border-radius: 4px;
  font-size: 0.875rem;
  position: relative;
  background-color: #fee8e7;
  border: 1px solid #faa19b;
  color: #000;
  padding: 12px;
}

.c1 {
  -webkit-animation: lbWRkT 300ms forwards;
  animation: lbWRkT 300ms forwards;
}

<div>
  <div
    class="c0 c1"
    role="alert"
    type="danger"
  >
    Alert
  </div>
</div>
`;

exports[`<Alert /> Snapshots should match snapshot with all types 5`] = `
.c0 {
  border-radius: 4px;
  font-size: 0.875rem;
  position: relative;
  background-color: #eaf4f0;
  border: 1px solid #a8d2bf;
  color: #000;
  padding: 12px;
}

.c1 {
  -webkit-animation: lbWRkT 300ms forwards;
  animation: lbWRkT 300ms forwards;
}

<div>
  <div
    class="c0 c1"
    role="alert"
    type="success"
  >
    Alert
  </div>
</div>
`;

exports[`<Alert /> Snapshots should match snapshot with closable 1`] = `
.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border: 0;
  border-radius: 50%;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 5px;
  position: absolute;
  right: 10px;
  -webkit-transition: background-color 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out;
  top: 10px;
}

.c2:hover,
.c2:focus {
  background-color: rgba(0,0,0,0.1);
}

.c0 {
  border-radius: 4px;
  font-size: 0.875rem;
  position: relative;
  background-color: #fdfdfd;
  border: 1px solid #ccc;
  color: #000;
  padding: 12px;
}

.c1 {
  -webkit-animation: lbWRkT 300ms forwards;
  animation: lbWRkT 300ms forwards;
}

<div>
  <div
    class="c0 c1"
    role="alert"
    type="neutral"
  >
    <button
      aria-label="close"
      class="c2"
    >
      <svg
        height="10"
        version="1.1"
        viewBox="0 0 47.971 47.971"
        width="10"
        x="0px"
        xmlns="http://www.w3.org/2000/svg"
        y="0px"
      >
        <g>
          <path
            d="M28.228,23.986L47.092,5.122c1.172-1.171,1.172-3.071,0-4.242c-1.172-1.172-3.07-1.172-4.242,0L23.986,19.744L5.121,0.88 c-1.172-1.172-3.07-1.172-4.242,0c-1.172,1.171-1.172,3.071,0,4.242l18.865,18.864L0.879,42.85c-1.172,1.171-1.172,3.071,0,4.242 C1.465,47.677,2.233,47.97,3,47.97s1.535-0.293,2.121-0.879l18.865-18.864L42.85,47.091c0.586,0.586,1.354,0.879,2.121,0.879 s1.535-0.293,2.121-0.879c1.172-1.171,1.172-3.071,0-4.242L28.228,23.986z"
          />
        </g>
      </svg>
    </button>
    Alert
  </div>
</div>
`;

exports[`<Alert /> Snapshots should match snapshot with closable and closeText 1`] = `
.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border: 0;
  border-radius: 50%;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 5px;
  position: absolute;
  right: 10px;
  -webkit-transition: background-color 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out;
  top: 10px;
  border-radius: none;
  color: #666;
  padding: 0;
  top: 12px;
}

.c2:hover,
.c2:focus {
  background-color: rgba(0,0,0,0.1);
}

.c2:hover,
.c2:focus {
  background-color: transparent;
  color: #111;
}

.c0 {
  border-radius: 4px;
  font-size: 0.875rem;
  position: relative;
  background-color: #fdfdfd;
  border: 1px solid #ccc;
  color: #000;
  padding: 12px;
}

.c1 {
  -webkit-animation: lbWRkT 300ms forwards;
  animation: lbWRkT 300ms forwards;
}

<div>
  <div
    class="c0 c1"
    role="alert"
    type="neutral"
  >
    <button
      class="c2"
    >
      <span>
        Close Now
      </span>
    </button>
    Alert
  </div>
</div>
`;
